File: solutions/cpp/k-balanced_binary_tree.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <cstdlib>
#include <iostream>
#include <memory>
#include <utility>

#include "./Binary_tree_prototype.h"

using std::cout;
using std::endl;
using std::make_unique;
using std::pair;
using std::tie;
using std::unique_ptr;

pair<BinaryTreeNode<int>*, int> FindKUnbalancedNodeHelper(
    const unique_ptr<BinaryTreeNode<int>>&, int);

// @include
BinaryTreeNode<int>* FindKUnbalancedNode(
    const unique_ptr<BinaryTreeNode<int>>& tree, int k) {
  return FindKUnbalancedNodeHelper(tree, k).first;
}

// If there is any k-unbalanced node in tree, the first value of the return
// value is a k-unbalanced node; otherwise, null.  If the first is not null,
// the second value of the return value is the number of nodes in tree.
pair<BinaryTreeNode<int>*, int> FindKUnbalancedNodeHelper(
    const unique_ptr<BinaryTreeNode<int>>& tree, int k) {
  if (tree == nullptr) {
    return {nullptr, 0};  // Base case.
  }

  // Early return if left subtree is not k-balanced.
  BinaryTreeNode<int>* left_subtree_result;
  int left_subtree_node_num;
  tie(left_subtree_result, left_subtree_node_num) =
      FindKUnbalancedNodeHelper(tree->left, k);
  if (left_subtree_result != nullptr) {
    return {left_subtree_result, 0};
  }
  // Early return if right subtree is not k-balanced.
  BinaryTreeNode<int>* right_subtree_result;
  int right_subtree_node_num;
  tie(right_subtree_result, right_subtree_node_num) =
      FindKUnbalancedNodeHelper(tree->right, k);
  if (right_subtree_result != nullptr) {
    return {right_subtree_result, 0};
  }

  int node_num = left_subtree_node_num + right_subtree_node_num + 1;
  if (abs(left_subtree_node_num - right_subtree_node_num) > k) {
    // tree is not k-balanced but its children are k-balanced.
    return {tree.get(), node_num};
  }
  return {nullptr, node_num};
}
// @exclude

int main(int argc, char* argv[]) {
  //      3
  //    2   5
  //  1    4 6
  unique_ptr<BinaryTreeNode<int>> root = make_unique<BinaryTreeNode<int>>(
      BinaryTreeNode<int>{3, nullptr, nullptr});
  root->left = make_unique<BinaryTreeNode<int>>(
      BinaryTreeNode<int>{2, nullptr, nullptr});
  root->left->left = make_unique<BinaryTreeNode<int>>(
      BinaryTreeNode<int>{1, nullptr, nullptr});
  root->right = make_unique<BinaryTreeNode<int>>(
      BinaryTreeNode<int>{5, nullptr, nullptr});
  root->right->left = make_unique<BinaryTreeNode<int>>(
      BinaryTreeNode<int>{4, nullptr, nullptr});
  root->right->right = make_unique<BinaryTreeNode<int>>(
      BinaryTreeNode<int>{6, nullptr, nullptr});
  int k = 0;
  BinaryTreeNode<int>* ans(FindKUnbalancedNode(root, k));
  assert(ans->data == 2);
  if (ans) {
    cout << ans->data << endl;
  }
  return 0;
}

File: solutions/python/k_balanced_binary_tree.py
# k-balanced_binary_tree.cpp b4b3a70d8ab942579f85b4416f980d05831af969
from binary_tree_prototype import BinaryTreeNode


# @include
def find_k_unbalanced_node(tree, k):
    return find_k_unbalanced_node_helper(tree, k)[0]


# If there is any k-unbalanced node in tree, the first value of the return
# value is a k-unbalanced node; otherwise, null.  If the first is not null,
# the second value of the return value is the number of nodes in tree.
def find_k_unbalanced_node_helper(tree, k):
    if not tree:
        return (None, 0)  # Base case.

    # Early return if left subtree is not k-balanced.
    left_result = find_k_unbalanced_node_helper(tree.left, k)
    if left_result[0]:
        return (left_result[0], 0)
    # Early return if right subtree is not k-balanced.
    right_result = find_k_unbalanced_node_helper(tree.right, k)
    if right_result[0]:
        return (right_result[0], 0)

    node_num = left_result[1] + right_result[1] + 1
    if abs(left_result[1] - right_result[1]) > k:
        # tree is not k-balanced but its children are k-balanced.
        return (tree, node_num)
    return (None, node_num)


# @exclude


def main():
    #      3
    #    2   5
    #  1    4 6
    root = BinaryTreeNode(3)
    root.left = BinaryTreeNode(2)
    root.left.left = BinaryTreeNode(1)
    root.right = BinaryTreeNode(5)
    root.right.left = BinaryTreeNode(4)
    root.right.right = BinaryTreeNode(6)
    k = 0
    ans = find_k_unbalanced_node(root, k)
    assert ans.data == 2
    if ans:
        print(ans.data)


if __name__ == '__main__':
    main()
